drivers/Watchdog.cpp

```diff
@@ -17,92 +17,95 @@
 #ifdef DEVICE_WATCHDOG
 
 #include "Watchdog.h"
+#include "VirtualWatchdog.h"
 
-namespace mbed {
+#define MS_TO_US(x) ((x) * 1000) //macro to convert millisecond to microsecond
 
-Watchdog *Watchdog::_first = NULL;
+namespace mbed {
 
-Watchdog::Watchdog(uint32_t timeout, const char *const str): _name(str)
-{
-    _current_count = 0;
-    _is_initialized = false;
-    _next = NULL;
-    _max_timeout = timeout;
-}
+#if DEVICE_LPTICKER
+    /** Create singleton instance of LowPowerTicker for watchdog periodic call back of kick.
+     */
+    SingletonPtr<LowPowerTicker> Watchdog::_ticker;
+#else
+    /** Create singleton instance of Ticker for watchdog periodic call back of kick.
+     */
+    SingletonPtr<Ticker> Watchdog::_ticker;
+#endif
 
-void Watchdog::add_to_list()
-{
-    this->_next = _first;
-    _first = this;
-    _is_initialized =  true;
-}
+bool Watchdog::_running = false;
+static const uint32_t elapsed_ms = MBED_CONF_TARGET_WATCHDOG_TIMEOUT / 2;
 
-void Watchdog::start()
+bool Watchdog::start()
 {
-    MBED_ASSERT(!_is_initialized);
+    watchdog_status_t sts;
+    MBED_ASSERT(MBED_CONF_TARGET_WATCHDOG_TIMEOUT < get_max_timeout());
     core_util_critical_section_enter();
-    add_to_list();
+    if (_running) {
+        core_util_critical_section_exit();
+        return false;
+    }
+    watchdog_config_t config;
+    config.timeout_ms = MBED_CONF_TARGET_WATCHDOG_TIMEOUT;
+    sts = hal_watchdog_init(&config);
+    if (sts == WATCHDOG_STATUS_OK) {
+        _running = true;
+    }
     core_util_critical_section_exit();
+    if (_running) {
+        us_timestamp_t timeout = (MS_TO_US(((elapsed_ms <= 0) ? 1 : elapsed_ms)));
+        _ticker->attach_us(callback(&Watchdog::kick), timeout);
+    }
+    return _running;
 }
 
 
-void Watchdog::kick()
+bool Watchdog::stop()
 {
-    MBED_ASSERT(_is_initialized);
+    watchdog_status_t sts;
+    bool msts = true;
     core_util_critical_section_enter();
-    _current_count = 0;
+    if (_running) {
+        sts = hal_watchdog_stop();
+        if (sts != WATCHDOG_STATUS_OK) {
+            msts = false;
+        } else {
+            _ticker->detach();
+            _running = false;
+        }
+
+    } else {
+        msts = false;
+    }
     core_util_critical_section_exit();
+    return msts;
 }
 
-void Watchdog::stop()
+void Watchdog::kick()
 {
-    MBED_ASSERT(_is_initialized);
     core_util_critical_section_enter();
-    remove_from_list();
+    hal_watchdog_kick();
+    // VirtualWatchdog will access the watchdog process method to verify
+    // all registered users/threads in alive state */
+    VirtualWatchdog::process(((elapsed_ms <= 0) ? 1 : elapsed_ms));
```

Comment:
This can be moved to the `VirtualWatchdog` itself if it can `kick()` this directly from a ticker instantiated there.

Reply:
I probably understand the point now, checking how to decouple